src/planner/plannodes/nested_loop_join_plan_node.cpp

```diff
@@ -1,29 +1,73 @@
-#include "planner/plannodes/nested_loop_join_plan_node.h"
-
 #include <memory>
+#include <utility>
 #include <vector>
 
+#include "planner/plannodes/nested_loop_join_plan_node.h"
+
 namespace terrier::planner {
 
 common::hash_t NestedLoopJoinPlanNode::Hash() const {
   common::hash_t hash = AbstractJoinPlanNode::Hash();
-
-  // There is nothing else for us to do here!
-
+  hash = common::HashUtil::CombineHashInRange(hash, left_keys_.begin(), left_keys_.end());
+  hash = common::HashUtil::CombineHashInRange(hash, right_keys_.begin(), right_keys_.end());
   return hash;
 }
 
 bool NestedLoopJoinPlanNode::operator==(const AbstractPlanNode &rhs) const {
   // There is nothing else for us to do here! Go home! You're drunk!
-  return AbstractJoinPlanNode::operator==(rhs);
+  // Unfortunately, now there is something to do...
+  if (!AbstractJoinPlanNode::operator==(rhs)) return false;
+
+  auto &other = reinterpret_cast<const NestedLoopJoinPlanNode &>(rhs);
+  if (left_keys_ != other.left_keys_) return false;
+  return right_keys_ == other.right_keys_;
 }
 
-nlohmann::json NestedLoopJoinPlanNode::ToJson() const { return AbstractJoinPlanNode::ToJson(); }
+nlohmann::json NestedLoopJoinPlanNode::ToJson() const {
+  auto j = AbstractJoinPlanNode::ToJson();
+  std::vector<nlohmann::json> left_keys;
+  left_keys.reserve(left_keys_.size());
+  for (const auto &key : left_keys_) {
+    left_keys.emplace_back(key->ToJson());
+  }
+  j["left_keys"] = left_keys;
+  std::vector<nlohmann::json> right_keys;
+  right_keys.reserve(right_keys_.size());
+  for (const auto &key : right_keys_) {
+    right_keys.emplace_back(key->ToJson());
+  }
+  j["right_keys"] = right_keys;
+  return j;
```

Comment:
You could write a Json serializer for `ManagedPointers` that just calls `underlying_->ToJson`, that would severely reduce the logic here and in other plan nodes.